diff -Nur llvm-3.7.0.src/tools/clang/lib/AST/ItaniumMangle.cpp llvm-3.7.0.srcn/tools/clang/lib/AST/ItaniumMangle.cpp
--- llvm-3.7.0.src/tools/clang/lib/AST/ItaniumMangle.cpp	2015-12-06 18:25:43.760217835 +0800
+++ llvm-3.7.0.srcn/tools/clang/lib/AST/ItaniumMangle.cpp	2015-12-06 18:26:40.385216821 +0800
@@ -651,7 +651,7 @@
 
 void CXXNameMangler::writeAbiTags(const NamedDecl *ND, const AbiTagList *AdditionalAbiTags) {
   assert(AbiTags && "require AbiTagState");
-  if (AbiTags) AbiTags->write(Out, ND, DisableDerivedAbiTags ? nullptr : AdditionalAbiTags);
+  if (AbiTags && ND) AbiTags->write(Out, ND, DisableDerivedAbiTags ? nullptr : AdditionalAbiTags);
 }
 
 void CXXNameMangler::mangle(const NamedDecl *D) {
