diff -Nur llvm-3.9.0.src/projects/CMakeLists.txt llvm-3.9.0.srcn/projects/CMakeLists.txt
--- llvm-3.9.0.src/projects/CMakeLists.txt	2015-10-29 02:36:56.000000000 +0800
+++ llvm-3.9.0.srcn/projects/CMakeLists.txt	2016-11-17 15:16:30.390897753 +0800
@@ -9,6 +9,7 @@
        (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/libcxx) AND
        (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/libcxxabi) AND
        (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/libunwind) AND
+       (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/openmp) AND
        (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/test-suite))
       add_subdirectory(${entry})
     endif()
@@ -26,6 +27,7 @@
     add_llvm_external_project(libcxxabi)
     add_llvm_external_project(libcxx)
     add_llvm_external_project(libunwind)
+    add_llvm_external_project(openmp)
   endif()
   if(NOT LLVM_BUILD_EXTERNAL_COMPILER_RT)
     add_llvm_external_project(compiler-rt)
