diff -Nur llvm-3.7.1.src/projects/CMakeLists.txt llvm-3.7.1.srcn/projects/CMakeLists.txt
--- llvm-3.7.1.src/projects/CMakeLists.txt	2015-10-17 10:23:42.000000000 +0800
+++ llvm-3.7.1.srcn/projects/CMakeLists.txt	2015-12-11 13:58:07.046779389 +0800
@@ -25,6 +26,7 @@
     add_llvm_external_project(libcxxabi)
     add_llvm_external_project(libcxx)
     add_llvm_external_project(libunwind)
+    add_llvm_external_project(openmp)
   endif()
   if(NOT LLVM_BUILD_EXTERNAL_COMPILER_RT)
     add_llvm_external_project(compiler-rt)
