diff -Nur glibc-2.15/locale/programs/locarchive.c glibc-2.15n/locale/programs/locarchive.c
--- glibc-2.15/locale/programs/locarchive.c	2012-01-01 20:16:32.000000000 +0800
+++ glibc-2.15n/locale/programs/locarchive.c	2012-04-10 21:59:08.187219279 +0800
@@ -253,7 +253,7 @@
 /* forward decls for below */
 static uint32_t add_locale (struct locarhandle *ah, const char *name,
 			    locale_data_t data, bool replace);
-static void add_alias (struct locarhandle *ah, const char *alias,
+void add_alias (struct locarhandle *ah, const char *alias,
 		       bool replace, const char *oldname,
 		       uint32_t *locrec_offset_p);
 
@@ -636,7 +636,7 @@
 #include "../../intl/explodename.c"
 #include "../../intl/l10nflist.c"
 
-static struct namehashent *
+struct namehashent *
 insert_name (struct locarhandle *ah,
 	     const char *name, size_t name_len, bool replace)
 {
@@ -694,7 +694,7 @@
   return &namehashtab[idx];
 }
 
-static void
+void
 add_alias (struct locarhandle *ah, const char *alias, bool replace,
 	   const char *oldname, uint32_t *locrec_offset_p)
 {
diff -Nur glibc-2.15/localedata/locales/zh_TW glibc-2.15n/localedata/locales/zh_TW
--- glibc-2.15/localedata/locales/zh_TW	2012-01-01 20:16:32.000000000 +0800
+++ glibc-2.15n/localedata/locales/zh_TW	2012-04-10 21:59:47.523219882 +0800
@@ -1,12 +1,12 @@
 comment_char %
 escape_char /
 %
-% Chinese language locale for Taiwan R.O.C.
+% Chinese language locale for Taiwan
 % charmap: BIG5-CP950
 %
 % Original Author:
 %    Ming-Che Chuang
-% 	PPE of NTU, Taiwan, ROC
+% 	PPE of NTU, Taiwan
 %
 % Currently Maintained by:
 %    Yuan-Chung Cheng
@@ -17,7 +17,7 @@
 % Reference:	http://wwwold.dkuug.dk/JTC1/SC22/WG20/docs/n690.pdf
 
 LC_IDENTIFICATION
-title      "Chinese locale for Taiwan R.O.C."
+title      "Chinese locale for Taiwan"
 source     ""
 address    ""
 contact    ""
@@ -25,7 +25,7 @@
 tel        ""
 fax        ""
 language   "Chinese"
-territory  "Taiwan R.O.C."
+territory  "Taiwan"
 revision   "0.2"
 date       "2000-08-02"
 %
diff -Nur glibc-2.15/localedata/Makefile glibc-2.15n/localedata/Makefile
--- glibc-2.15/localedata/Makefile	2012-01-01 20:16:32.000000000 +0800
+++ glibc-2.15n/localedata/Makefile	2012-04-10 22:00:24.406220421 +0800
@@ -224,6 +224,7 @@
 	echo -n '...'; \
 	input=`echo $$locale | sed 's/\([^.]*\)[^@]*\(.*\)/\1\2/'`; \
 	$(LOCALEDEF) --alias-file=../intl/locale.alias \
+			 --no-archive \
 		     -i locales/$$input -c -f charmaps/$$charset \
 		     $(addprefix --prefix=,$(install_root)) $$locale; \
 	echo ' done'; \
