diff -Nur docbook-xsl-1.79.1/manpages/other.xsl docbook-xsl-1.79.1n/manpages/other.xsl
--- docbook-xsl-1.79.1/manpages/other.xsl	2014-11-21 07:00:06.000000000 +0800
+++ docbook-xsl-1.79.1n/manpages/other.xsl	2016-12-07 12:17:53.205253465 +0800
@@ -560,7 +560,14 @@
           <xsl:with-param name="message-prolog">Note: </xsl:with-param>
           <xsl:with-param name="message-epilog"> (soelim stub)</xsl:with-param>
           <xsl:with-param name="content">
-            <xsl:value-of select="'.so '"/>
+            <xsl:choose>
+              <xsl:when test="$man.output.in.separate.dir = 0">
+              <xsl:value-of select="concat('.so man', $section, '/')"/>
+              </xsl:when>
+              <xsl:otherwise>
+              <xsl:value-of select="'.so '"/> <!-- added case -->
+              </xsl:otherwise>
+              </xsl:choose>
             <xsl:variable name="full.filename">
               <xsl:call-template name="make.adjusted.man.filename">
                 <xsl:with-param name="name" select="$first.refname"/>
