--- man-1.6gn/src/Makefile.in	2014-10-23 14:35:01.231358002 -0800
+++ man-1.6gnew/src/Makefile.in	2014-10-23 14:37:20.681815371 -0800
@@ -102,7 +102,7 @@
 
 install: all apropos whatis makewhatis
 	mkdir -p $(DESTDIR)$(PREFIX)@bindir@
-	$(INSTALL) -c @man_install_flags@ man$(EXEEXT) $(DESTDIR)$(PREFIX)@man@
+	$(INSTALL) -c -m 755 man$(EXEEXT) $(DESTDIR)$(PREFIX)@man@
 	$(INSTALL) -c -m 755 apropos $(DESTDIR)$(PREFIX)@apropos@
 	$(INSTALL) -c -m 755 whatis $(DESTDIR)$(PREFIX)@whatis@
 	$(INSTALL) -c -m 755 man2dvi $(DESTDIR)$(PREFIX)@man2dvi@
