diff -Nur git-2.0.0.rc4/contrib/hooks/update-paranoid git-2.0.0.rc4n/contrib/hooks/update-paranoid
--- git-2.0.0.rc4/contrib/hooks/update-paranoid	2014-05-21 07:17:49.000000000 +0800
+++ git-2.0.0.rc4n/contrib/hooks/update-paranoid	2014-06-27 19:06:09.640172665 +0800
@@ -86,8 +86,8 @@
 objects/info/alternates file lists this repository and all other
 possible sources, and whose refs subdirectory contains symlinks
 to this repository's refs subdirectory, and to all other possible
-sources refs subdirectories.  Yes, this means that you cannot
-use packed-refs in those repositories as they won't be resolved
+sources refs subdirectories.  Yes, this means that you cannot use 
+packed-refs in those repositories as they won't be resolved
 correctly.
 
 =cut
