diff --git a/lib/rpmdb.c b/lib/rpmdb.c
index 5638b84..2c5db09 100644
--- a/lib/rpmdb.c
+++ b/lib/rpmdb.c
@@ -804,6 +804,27 @@ int rpmdbCountPackages(rpmdb db, const char * name)
     return count;
 }
 
+int rpmdbCountPackagesISoftApp(rpmdb db, const char * name)
+{
+    int count = -1;
+    dbiIndex dbi = NULL;
+
+    if (name != NULL && pkgdbOpen(db, 0, &dbi) == 0) {
+        dbiIndexSet matches = NULL;
+
+	    rpmRC rc = indexGet(dbi, name, strlen(name), &matches);
+
+	    if (rc == RPMRC_OK) {
+	        count = dbiIndexSetCount(matches);
+	    } else {
+	        count = (rc == RPMRC_NOTFOUND) ? 0 : -1;
+	    }
+	    dbiIndexSetFree(matches);
+    }
+
+    return count;
+}
+
 int rpmdbCountProvides(rpmdb db, const char * name)
 {
     int count = -1;
diff --git a/lib/rpmdb.h b/lib/rpmdb.h
index c8ca0e9..f116435 100644
--- a/lib/rpmdb.h
+++ b/lib/rpmdb.h
@@ -61,6 +61,14 @@ int rpmdbOpenAll (rpmdb db);
  */
 int rpmdbCountPackages(rpmdb db, const char * name);
 
+/** \ingroup rpmdb
+ * Return number of instances of Packages in rpm or isoft-app database.
+ * @param db		rpm or isoft-app database
+ * @param name		rpm or isoft-app package name
+ * @return		number of instances
+ */
+int rpmdbCountPackagesISoftApp(rpmdb db, const char * name);
+
 /** \ingroupd rpmdb
  * Return number of instances of provides in rpm or isoft-app database.
  * @param db        rpm or isoft-app database
