diff -Nur rpm-4.9.1.2/macros.in rpm-4.9.1.2n/macros.in
--- rpm-4.9.1.2/macros.in	2012-03-08 15:43:11.659078545 +0800
+++ rpm-4.9.1.2n/macros.in	2012-03-08 15:42:52.420078249 +0800
@@ -374,13 +374,13 @@
 # Should unpackaged files in a build root terminate a build?
 #
 # Note: The default value should be 0 for legacy compatibility.
-%_unpackaged_files_terminate_build	1
+%_unpackaged_files_terminate_build 0	
 
 #
 # Should missing %doc files in the build directory terminate a build?
 #
 # Note: The default value should be 0 for legacy compatibility.
-%_missing_doc_files_terminate_build	1
+%_missing_doc_files_terminate_build 0	
 
 #
 # Should binaries in noarch packages terminate a build?
