diff -Nur rpm-4.12.0/macros.in rpm-4.12.0n/macros.in
--- rpm-4.12.0/macros.in	2014-10-20 15:21:56.808941168 +0000
+++ rpm-4.12.0n/macros.in	2014-10-20 15:22:51.028936920 +0000
@@ -60,6 +60,7 @@
 %__patch		@__PATCH@
 %__perl			@__PERL@
 %__python		@__PYTHON@
+%__python3      /usr/bin/python3
 %__restorecon		@__RESTORECON@
 %__rm			@__RM@
 %__rsh			@__RSH@
@@ -991,6 +992,10 @@
 %python_sitearch %(%{__python} -c "from distutils.sysconfig import get_python_lib; import sys; sys.stdout.write(get_python_lib(1))")
 %python_version %(%{__python} -c "import sys; sys.stdout.write(sys.version[:3])")
 
+%python3_sitelib %(%{__python3} -c "from distutils.sysconfig import get_python_lib; import sys; sys.stdout.write(get_python_lib())")
+%python3_sitearch %(%{__python3} -c "from distutils.sysconfig import get_python_lib; import sys; sys.stdout.write(get_python_lib(1))")
+%python3_version %(%{__python3} -c "import sys; sys.stdout.write(sys.version[:3])")
+
 #------------------------------------------------------------------------------
 # arch macro for all Intel i?86 compatibile processors
 #  (Note: This macro (and it's analogues) will probably be obsoleted when
