diff -Nur rpm-4.13.0-rc1/macros.in rpm-4.13.0-rc1n/macros.in
--- rpm-4.13.0-rc1/macros.in	2015-09-25 13:48:41.099619602 +0800
+++ rpm-4.13.0-rc1n/macros.in	2015-09-25 13:49:29.636647933 +0800
@@ -381,13 +381,13 @@
 # Should unpackaged files in a build root terminate a build?
 #
 # Note: The default value should be 0 for legacy compatibility.
-%_unpackaged_files_terminate_build	1
+%_unpackaged_files_terminate_build	0	
 
 #
 # Should missing %doc files in the build directory terminate a build?
 #
 # Note: The default value should be 0 for legacy compatibility.
-%_missing_doc_files_terminate_build	1
+%_missing_doc_files_terminate_build	0
 
 #
 # Should empty %files manifest file terminate a build?
