diff -Nur os-prober/os-probes/mounted/common/90linux-distro os-probern/os-probes/mounted/common/90linux-distro
--- os-prober/os-probes/mounted/common/90linux-distro	2012-03-07 07:56:26.306010388 +0800
+++ os-probern/os-probes/mounted/common/90linux-distro	2012-03-07 08:00:18.635013945 +0800
@@ -121,6 +121,10 @@
 	elif [ -e "$dir/etc/meego-release" ]; then
 		short="MeeGo"
 		long="$(head -1 "$dir/etc/meego-release")"
+	elif [ -e "$dir/etc/os-release" ]; then
+       . /etc/os-release
+       short=$NAME
+       long=$PRETTY_NAME
 	else
 		short="Linux"
 		long="unknown Linux distribution"
