diff -Nur bzip2-1.0.6/Makefile-libbz2_so bzip2-1.0.6n/Makefile-libbz2_so
--- bzip2-1.0.6/Makefile-libbz2_so	2010-09-10 23:07:52.000000000 +0000
+++ bzip2-1.0.6n/Makefile-libbz2_so	2015-06-19 07:58:10.185204747 +0000
@@ -35,7 +35,7 @@
       bzlib.o
 
 all: $(OBJS)
-	$(CC) -shared -Wl,-soname -Wl,libbz2.so.1.0 -o libbz2.so.1.0.6 $(OBJS)
+	$(CC) -shared -Wl,-soname -Wl,libbz2.so.1 -o libbz2.so.1.0.6 $(OBJS)
 	$(CC) $(CFLAGS) -o bzip2-shared bzip2.c libbz2.so.1.0.6
 	rm -f libbz2.so.1.0
 	ln -s libbz2.so.1.0.6 libbz2.so.1.0
