From 84039362e4da04b863cdc8995d415621b4b3100a Mon Sep 17 00:00:00 2001
From: sulit <sulitsrc@gmail.com>
Date: Tue, 17 Nov 2015 17:13:25 +0800
Subject: [PATCH 83/83] fixed logical error

---
 util/grub-mkconfig.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/util/grub-mkconfig.in b/util/grub-mkconfig.in
index 34ce9af..85c1066 100644
--- a/util/grub-mkconfig.in
+++ b/util/grub-mkconfig.in
@@ -154,7 +154,7 @@ fi
 probe_othersystem=$(mktemp)
 ${grub_mkconfig_dir}/30* &> $probe_othersystem
 grep Found $probe_othersystem &> /dev/null && grep menuentry $probe_othersystem &> /dev/null
-if [ $? ]; then
+if [ ! $? ]; then
         if [ -z "$GRUB_TIMEOUT" ]; then
                 GRUB_TIMEOUT=0
         fi
-- 
2.6.3

