diff -purN thin-provisioning-tools-0.6.0.orig/Makefile.in thin-provisioning-tools-0.6.0.new/Makefile.in
--- thin-provisioning-tools-0.6.0.orig/Makefile.in	2016-01-20 16:43:21.000000000 +0100
+++ thin-provisioning-tools-0.6.0.new/Makefile.in	2016-01-20 17:26:11.943664295 +0100
@@ -174,7 +174,6 @@ distclean: clean
 install: bin/pdata_tools
 	$(INSTALL_DIR) $(BINDIR)
 	$(INSTALL_PROGRAM) bin/pdata_tools $(BINDIR)
-	$(STRIP) $(BINDIR)/pdata_tools
 	ln -s -f pdata_tools $(BINDIR)/cache_check
 	ln -s -f pdata_tools $(BINDIR)/cache_dump
 	ln -s -f pdata_tools $(BINDIR)/cache_metadata_size
