diff -Nur libobjc2-1.6.1/Makefile libobjc2-1.6.1n/Makefile
--- libobjc2-1.6.1/Makefile	2012-03-25 00:43:01.000000000 +0800
+++ libobjc2-1.6.1n/Makefile	2012-11-26 17:00:53.693125481 +0800
@@ -73,13 +73,13 @@
 
 $(LIBOBJCXX).so.$(VERSION): $(LIBOBJC).so.$(VERSION) $(OBJCXX_OBJECTS)
 	$(SILENT)echo Linking shared Objective-C++ runtime library...
-	$(SILENT)$(CXX) -shared \
+	$(SILENT)$(CXX) -shared -ldispatch \
             -Wl,-soname=$(LIBOBJCXX).so.$(MAJOR_VERSION) \
             -o $@ $(OBJCXX_OBJECTS)
 
 $(LIBOBJC).so.$(VERSION): $(OBJECTS)
 	$(SILENT)echo Linking shared Objective-C runtime library...
-	$(SILENT)$(CC) -shared -rdynamic \
+	$(SILENT)$(CC) -shared -rdynamic -ldispatch \
             -Wl,-soname=$(LIBOBJC).so.$(MAJOR_VERSION) \
             -o $@ $(OBJECTS)
 
