diff -Nur fontconfig-2.11.1/conf.d/65-nonlatin.conf fontconfig-2.11.1n/conf.d/65-nonlatin.conf
--- fontconfig-2.11.1/conf.d/65-nonlatin.conf	2014-07-17 09:21:00.524983753 +0800
+++ fontconfig-2.11.1n/conf.d/65-nonlatin.conf	2014-07-17 09:22:38.194980261 +0800
@@ -25,6 +25,7 @@
 			<family>Hapax Berbère</family> <!-- tifinagh -->
 			<family>MS Mincho</family> <!-- han (ja) -->
 			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>Source Han Sans CN</family> <!-- han (zh-cn,zh-tw) -->
 			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>SimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>PMingLiu</family> <!-- han (zh-tw) -->
@@ -100,6 +101,7 @@
 			<family>UmePlus P Gothic</family> <!-- han (ja) -->
 			<!-- chinese fonts are actually serifed -->
 			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>Source Han Sans CN</family> <!-- han (zh-cn,zh-tw) -->
 			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>SimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>PMingLiu</family> <!-- han (zh-tw) -->
@@ -156,6 +158,7 @@
 			<family>MS Gothic</family> <!-- han (ja) -->
 			<family>UmePlus Gothic</family> <!-- han (ja) -->
 			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>Source Han Sans CN</family> <!-- han (zh-cn,zh-tw) -->
 			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>NSimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>MingLiu</family> <!-- han (zh-tw) -->
