diff -Nur fontconfig-2.11.0/conf.d/65-nonlatin.conf fontconfig-2.11.0n/conf.d/65-nonlatin.conf
--- fontconfig-2.11.0/conf.d/65-nonlatin.conf	2013-10-19 14:28:07.155302406 +0800
+++ fontconfig-2.11.0n/conf.d/65-nonlatin.conf	2013-10-19 14:28:52.503300785 +0800
@@ -24,6 +24,8 @@
 			<family>padmaa</family> <!-- gujarati -->
 			<family>Hapax Berbère</family> <!-- tifinagh -->
 			<family>MS Mincho</family> <!-- han (ja) -->
+			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>SimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>PMingLiu</family> <!-- han (zh-tw) -->
 			<family>WenQuanYi Zen Hei</family> <!-- han (zh-cn,zh-tw) -->
@@ -97,6 +99,8 @@
 			<family>MS Gothic</family> <!-- han (ja) -->
 			<family>UmePlus P Gothic</family> <!-- han (ja) -->
 			<!-- chinese fonts are actually serifed -->
+			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>SimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>PMingLiu</family> <!-- han (zh-tw) -->
 			<family>WenQuanYi Zen Hei</family> <!-- han (zh-cn,zh-tw) -->
@@ -151,6 +154,8 @@
 			<family>AR PL KaitiM GB</family>
 			<family>MS Gothic</family> <!-- han (ja) -->
 			<family>UmePlus Gothic</family> <!-- han (ja) -->
+			<family>Microsoft YaHei</family> <!-- han (zh-cn,zh-tw) -->
+			<family>WenQuanYi Micro Hei</family> <!-- han (zh-cn,zh-tw) -->
 			<family>NSimSun</family> <!-- han (zh-cn,zh-tw) -->
 			<family>MingLiu</family> <!-- han (zh-tw) -->
 			<family>AR PL ShanHeiSun Uni</family> <!-- han (ja,zh-cn,zh-tw) -->
