diff -Nur valgrind-3.10.1/config.h.in valgrind-3.10.1n/config.h.in
--- valgrind-3.10.1/config.h.in	2014-11-25 19:42:18.000000000 +0000
+++ valgrind-3.10.1n/config.h.in	2015-06-20 08:06:48.578448090 +0000
@@ -69,6 +69,9 @@
 /* Define to 1 if you're using glibc 2.20.x */
 #undef GLIBC_2_20
 
+/* Define to 1 if you're using glibc 2.21.x */
+#undef GLIBC_2_21
+
 /* Define to 1 if you're using glibc 2.3.x */
 #undef GLIBC_2_3
 
diff -Nur valgrind-3.10.1/configure.ac valgrind-3.10.1n/configure.ac
--- valgrind-3.10.1/configure.ac	2014-11-25 19:41:13.000000000 +0000
+++ valgrind-3.10.1n/configure.ac	2015-06-20 08:07:31.746446919 +0000
@@ -1018,6 +1018,13 @@
 	DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
 	DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
 	;;
+     2.21)
+    AC_MSG_RESULT(2.21 family)
+    AC_DEFINE([GLIBC_2_21], 1, [Define to 1 if you're using glibc 2.21.x])
+    DEFAULT_SUPP="glibc-2.X.supp ${DEFAULT_SUPP}"
+    DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
+    DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
+    ;;
      darwin)
 	AC_MSG_RESULT(Darwin)
 	AC_DEFINE([DARWIN_LIBC], 1, [Define to 1 if you're using Darwin])
@@ -1031,7 +1038,7 @@
 
      *)
 	AC_MSG_RESULT([unsupported version ${GLIBC_VERSION}])
-	AC_MSG_ERROR([Valgrind requires glibc version 2.2 - 2.19])
+	AC_MSG_ERROR([Valgrind requires glibc version 2.2 - 2.21])
 	AC_MSG_ERROR([or Darwin or Bionic libc])
 	;;
 esac
