diff -Nur nss-util-3.27.1/nss/cmd/manifest.mn nss-util-3.27.1n/nss/cmd/manifest.mn
--- nss-util-3.27.1/nss/cmd/manifest.mn	2016-12-29 11:14:43.000000000 +0800
+++ nss-util-3.27.1n/nss/cmd/manifest.mn	2016-12-29 11:30:35.433491410 +0800
@@ -33,48 +33,6 @@
 ifndef NSS_BUILD_SOFTOKEN_ONLY
 ifndef NSS_BUILD_UTIL_ONLY
 NSS_SRCDIRS = \
- addbuiltin \
- atob  \
- btoa  \
- certcgi \
- certutil  \
- chktest  \
- crlutil  \
- crmftest \
- dbtest \
- derdump  \
- digest  \
- httpserv  \
- listsuites \
- makepqg  \
- multinit \
- ocspclnt  \
- ocspresp \
- oidcalc  \
- p7content  \
- p7env  \
- p7sign  \
- p7verify  \
- pk12util \
- pk11gcmtest \
- pk11mode \
- pk1sign  \
- pp  \
- pwdecrypt \
- rsaperf \
- sdrtest \
- selfserv  \
- signtool \
- signver \
- smimetools  \
- ssltap  \
- strsclnt \
- symkeyutil \
- tests \
- tstclnt  \
- vfychain \
- vfyserv \
- modutil \
  $(NULL)
 
 ifndef NSS_DISABLE_LIBPKIX
diff -Nur nss-util-3.27.1/nss/lib/manifest.mn nss-util-3.27.1n/nss/lib/manifest.mn
--- nss-util-3.27.1/nss/lib/manifest.mn	2016-12-29 11:14:43.000000000 +0800
+++ nss-util-3.27.1n/nss/lib/manifest.mn	2016-12-29 11:27:45.783498974 +0800
@@ -50,8 +50,6 @@
 #  crmf jar (not dll's)
 DIRS = \
 	$(UTIL_SRCDIR) \
-	$(SOFTOKEN_SRCDIRS) \
-	$(NSS_SRCDIRS) \
 	$(NULL)
 
 #  fortcrypt  is no longer built
