--- linux-4.3.0.x86_64/drivers/platform/x86/ideapad-laptop.c	2015-11-15 11:47:48.502878305 +0800
+++ linux-4.3.0.x86_64n/drivers/platform/x86/ideapad-laptop.c	2015-11-15 11:51:49.998873980 +0800
@@ -898,6 +898,13 @@
 			DMI_MATCH(DMI_PRODUCT_VERSION, "Lenovo YOGA 3"),
 		},
 	},
+	{
+	        .ident = "Lenovo Yoga 900",
+	        .matches = {
+	                DMI_MATCH(DMI_SYS_VENDOR, "LENOVO"),
+	                DMI_MATCH(DMI_PRODUCT_VERSION, "Lenovo YOGA 900"),
+	        },
+	},
 	{}
 };
 
