diff -Nur linux-4.3.0.x86_64/./drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c linux-4.3.0.x86_64n/./drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c
--- linux-4.3.0.x86_64/./drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c	2015-11-13 10:10:23.000000000 +0800
+++ linux-4.3.0.x86_64n/./drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c	2015-11-13 10:53:39.256863876 +0800
@@ -683,9 +683,10 @@
 	crtc = &adev->mode_info.crtcs[pipe]->base;
 
 	/* Helper routine in DRM core does all the work: */
-	return drm_calc_vbltimestamp_from_scanoutpos(dev, pipe, max_error,
-						     vblank_time, flags,
-						     &crtc->hwmode);
+        return drm_calc_vbltimestamp_from_scanoutpos(dev, pipe, max_error,
+                                                     vblank_time, flags,
+                                                     crtc, &crtc->hwmode);
+
 }
 
 const struct drm_ioctl_desc amdgpu_ioctls_kms[] = {
