--- linux-4.3.0.x86_64/drivers/platform/x86/asus-wmi.c	2015-11-02 08:05:25.000000000 +0800
+++ linux-4.3.0.x86_64n/drivers/platform/x86/asus-wmi.c	2015-11-15 11:48:46.689877263 +0800
@@ -1680,7 +1680,7 @@
 	int rv, err, value;
 
 	value = asus_wmi_get_devstate_simple(asus, devid);
-	if (value == -ENODEV)	/* Check device presence */
+	if (value < 0)	/* Check device presence */
 		return value;
 
 	rv = parse_arg(buf, count, &value);
