--- linux-4.4.0/drivers/gpu/drm/i915/intel_opregion.c	2015-12-07 07:43:12.000000000 +0800
+++ linux-4.4.0n/drivers/gpu/drm/i915/intel_opregion.c	2015-12-08 15:37:37.458476695 +0800
@@ -25,6 +25,8 @@
  *
  */
 
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
 #include <linux/acpi.h>
 #include <acpi/video.h>
 
@@ -682,7 +684,7 @@
 	}
 
 	if (!acpi_video_bus) {
-		DRM_ERROR("No ACPI video bus found\n");
+		pr_warn("No ACPI video bus found\n");
 		return;
 	}
 
