diff -Nur linux-4.3.0.x86_64/drivers/gpu/drm/Kconfig linux-4.3.0.x86_64n/drivers/gpu/drm/Kconfig
--- linux-4.3.0.x86_64/drivers/gpu/drm/Kconfig	2015-11-02 08:05:25.000000000 +0800
+++ linux-4.3.0.x86_64n/drivers/gpu/drm/Kconfig	2015-11-13 10:19:16.423900817 +0800
@@ -160,6 +160,7 @@
 	  If M is selected, the module will be called amdgpu.
 
 source "drivers/gpu/drm/amd/amdgpu/Kconfig"
+source "drivers/gpu/drm/amd/powerplay/Kconfig"
 
 source "drivers/gpu/drm/nouveau/Kconfig"
 
