--- linux-4.3.0.x86_64/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/gk20a.c	2015-11-02 08:05:25.000000000 +0800
+++ linux-4.3.0.x86_64n/drivers/gpu/drm/nouveau/nvkm/subdev/pmu/gk20a.c	2015-11-09 18:16:20.666384086 +0800
@@ -202,6 +202,10 @@
 	.p_smooth = 1,
 };
 
+static const struct nvkm_pmu_func
+gk20a_pmu_func = {
+};
+
 static const struct nvkm_subdev_func
 gk20a_pmu = {
 	.init = gk20a_pmu_init,
@@ -221,6 +225,7 @@
 	*ppmu = &pmu->base;
 
 	nvkm_subdev_ctor(&gk20a_pmu, device, index, 0, &pmu->base.subdev);
+        pmu->base.func = &gk20a_pmu_func;
 	pmu->data = &gk20a_dvfs_data;
 	nvkm_alarm_init(&pmu->alarm, gk20a_pmu_dvfs_work);
 	return 0;
