--- linux-4.3.0.x86_64/drivers/platform/x86/thinkpad_acpi.c	2015-11-02 08:05:25.000000000 +0800
+++ linux-4.3.0.x86_64n/drivers/platform/x86/thinkpad_acpi.c	2015-11-15 11:54:54.660870673 +0800
@@ -6459,8 +6459,7 @@
 		pr_info("detected a 8-level brightness capable ThinkPad\n");
 		break;
 	default:
-		pr_err("Unsupported brightness interface, "
-		       "please contact %s\n", TPACPI_MAIL);
+		pr_info("Unsupported brightness interface\n");
 		tp_features.bright_unkfw = 1;
 		bright_maxlvl = b - 1;
 	}
