diff -Nur libxml2-2.9.2/parser.c libxml2-2.9.2n/parser.c
--- libxml2-2.9.2/parser.c	2014-10-16 05:57:40.000000000 +0000
+++ libxml2-2.9.2n/parser.c	2015-06-20 04:04:14.390843060 +0000
@@ -7235,7 +7235,8 @@
      * far more secure as the parser will only process data coming from
      * the document entity by default.
      */
-    if ((ent->checked == 0) &&
+    if (((ent->checked == 0) ||
+          ((ent->children == NULL) && (ctxt->options & XML_PARSE_NOENT))) &&
         ((ent->etype != XML_EXTERNAL_GENERAL_PARSED_ENTITY) ||
          (ctxt->options & (XML_PARSE_NOENT | XML_PARSE_DTDVALID)))) {
 	unsigned long oldnbent = ctxt->nbentities;
@@ -14831,7 +14832,6 @@
 	xmlXPathInit();
 #endif
 #ifdef LIBXML_CATALOG_ENABLED
-        xmlInitializeCatalog();
 #endif
 	xmlParserInitialized = 1;
 #ifdef LIBXML_THREAD_ENABLED
