From 11f1d4c897c8d015f689527f85ad94804ad15307 Mon Sep 17 00:00:00 2001
From: Jan Kaluza <jkaluza@redhat.com>
Date: Tue, 4 Dec 2012 12:07:36 +0100
Subject: [PATCH] file-5.04-generic-msdos.patch

Signed-off-by: Kamil Dudka <kdudka@redhat.com>
---
 magic/Magdir/msdos | 38 +++++++++++++++++++-------------------
 1 file changed, 19 insertions(+), 19 deletions(-)

diff --git a/magic/Magdir/msdos b/magic/Magdir/msdos
index d1a19a6..9e89f11 100644
--- a/magic/Magdir/msdos
+++ b/magic/Magdir/msdos
@@ -404,7 +404,7 @@
 # GRR: line below too general as it catches also 
 # rt.lib DYADISKS.PIC and many more
 # start with assembler instruction MOV
-0	ubyte		0x8c		
+#0	ubyte		0x8c		
 # skip "AppleWorks word processor data" like ARTICLE.1 ./apple
 >4	string			!O====	
 # skip some unknown basic binaries like RocketRnger.SHR
@@ -428,17 +428,17 @@
 # updated by Joerg Jenderek at Oct 2008
 0	ulelong		0xffff10eb	DR-DOS executable (COM)
 # byte 0xeb conflicts with "sequent" magic leshort 0xn2eb
-0	ubeshort&0xeb8d	>0xeb00		
+#0	ubeshort&0xeb8d	>0xeb00		
 # DR-DOS STACKER.COM SCREATE.SYS missed
 
 0       name    msdos-com
 >0  byte        x               DOS executable (COM)
 >6	string		SFX\ of\ LHarc	\b, %s
->0x1FE leshort	0xAA55		    \b, boot code
->85	string		UPX		        \b, UPX compressed
->4	string		\ $ARX		    \b, ARX self-extracting archive
->4	string		\ $LHarc	    \b, LHarc self-extracting archive
->0x20e string	SFX\ by\ LARC	\b, LARC self-extracting archive
+#>0x1FE leshort	0xAA55		    \b, boot code
+#>85	string		UPX		        \b, UPX compressed
+#>4	string		\ $ARX		    \b, ARX self-extracting archive
+#>4	string		\ $LHarc	    \b, LHarc self-extracting archive
+#>0x20e string	SFX\ by\ LARC	\b, LARC self-extracting archive
 
 # JMP 8bit
 0	        byte	0xeb
@@ -508,27 +508,27 @@
 # GRR search is not working
 #2	search/28	\xcd\x21	COM executable for MS-DOS
 #WHICHFAT.cOM
-2	string	\xcd\x21		COM executable for DOS
+#2	string	\xcd\x21		COM executable for DOS
 #DELTREE.cOM DELTREE2.cOM
-4	string	\xcd\x21		COM executable for DOS
+#4	string	\xcd\x21		COM executable for DOS
 #IFMEMDSK.cOM ASSIGN.cOM COMP.cOM
-5	string	\xcd\x21		COM executable for DOS
+#5	string	\xcd\x21		COM executable for DOS
 #DELTMP.COm HASFAT32.cOM
-7	string	\xcd\x21		
->0	byte	!0xb8			COM executable for DOS
+#7	string	\xcd\x21		
+#>0	byte	!0xb8			COM executable for DOS
 #COMP.cOM MORE.COm
-10	string	\xcd\x21		
->5	string	!\xcd\x21		COM executable for DOS
+#10	string	\xcd\x21		
+#>5	string	!\xcd\x21		COM executable for DOS
 #comecho.com
-13	string	\xcd\x21		COM executable for DOS
+#13	string	\xcd\x21		COM executable for DOS
 #HELP.COm EDIT.coM
-18	string	\xcd\x21		COM executable for MS-DOS
+#18	string	\xcd\x21		COM executable for MS-DOS
 #NWRPLTRM.COm
-23	string	\xcd\x21		COM executable for MS-DOS
+#23	string	\xcd\x21		COM executable for MS-DOS
 #LOADFIX.cOm LOADFIX.cOm
-30	string	\xcd\x21		COM executable for MS-DOS
+#30	string	\xcd\x21		COM executable for MS-DOS
 #syslinux.com 3.11
-70	string	\xcd\x21		COM executable for DOS
+#70	string	\xcd\x21		COM executable for DOS
 # many compressed/converted COMs start with a copy loop instead of a jump
 0x6	search/0xa	\xfc\x57\xf3\xa5\xc3	COM executable for MS-DOS
 0x6	search/0xa	\xfc\x57\xf3\xa4\xc3	COM executable for DOS
-- 
2.5.5

