diff -Nur xf86-video-virgl/src/compat-api.h xf86-video-virgln/src/compat-api.h
--- xf86-video-virgl/src/compat-api.h	2015-11-15 12:35:09.000000000 +0800
+++ xf86-video-virgln/src/compat-api.h	2016-11-29 13:07:10.792406110 +0800
@@ -75,8 +75,13 @@
 
 #define SCREEN_INIT_ARGS_DECL ScreenPtr pScreen, int argc, char **argv
 
+#if ABI_VIDEODRV_VERSION >= SET_ABI_VERSION(23, 0)
+#define BLOCKHANDLER_ARGS_DECL ScreenPtr arg, pointer pTimeout
+#define BLOCKHANDLER_ARGS arg, pTimeout
+#else
 #define BLOCKHANDLER_ARGS_DECL ScreenPtr arg, pointer pTimeout, pointer pReadmask
 #define BLOCKHANDLER_ARGS arg, pTimeout, pReadmask
+#endif
 
 #define CLOSE_SCREEN_ARGS_DECL ScreenPtr pScreen
 #define CLOSE_SCREEN_ARGS pScreen
diff -Nur xf86-video-virgl/src/uxa/uxa-damage.c xf86-video-virgln/src/uxa/uxa-damage.c
--- xf86-video-virgl/src/uxa/uxa-damage.c	2015-11-15 12:35:09.000000000 +0800
+++ xf86-video-virgln/src/uxa/uxa-damage.c	2016-11-29 13:07:14.728406324 +0800
@@ -28,6 +28,9 @@
 #include "uxa-priv.h"
 
 #include    <X11/X.h>
+#include    <X11/fonts/font.h>
+#include    <X11/fonts/fontstruct.h>
+#include    <X11/fonts/fontutil.h>
 #include    "scrnintstr.h"
 #include    "windowstr.h"
 #include    "dixfontstr.h"
