From eb36924ead40564325aa56d54a973dc8fb4eae83 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 24 Nov 2015 08:31:21 -0500
Subject: dix: Remove redundant ChangeWindowProperty

Use dixChangeWindowProperty(serverClient, ...) instead.

Signed-off-by: Adam Jackson <ajax@redhat.com>
Reviewed-by: Keith Packard <keithp@keithp.com>

diff --git a/dix/property.c b/dix/property.c
index 99608af..92c2558 100644
--- a/dix/property.c
+++ b/dix/property.c
@@ -355,14 +355,6 @@ dixChangeWindowProperty(ClientPtr pClient, WindowPtr pWin, Atom property,
 }
 
 int
-ChangeWindowProperty(WindowPtr pWin, Atom property, Atom type, int format,
-                     int mode, unsigned long len, void *value, Bool sendevent)
-{
-    return dixChangeWindowProperty(serverClient, pWin, property, type, format,
-                                   mode, len, value, sendevent);
-}
-
-int
 DeleteProperty(ClientPtr client, WindowPtr pWin, Atom propName)
 {
     PropertyPtr pProp, prevProp;
diff --git a/hw/xfree86/common/xf86Events.c b/hw/xfree86/common/xf86Events.c
index 6570f0b..709afd6 100644
--- a/hw/xfree86/common/xf86Events.c
+++ b/hw/xfree86/common/xf86Events.c
@@ -448,9 +448,10 @@ xf86UpdateHasVTProperty(Bool hasVT)
     if (property_name == BAD_RESOURCE)
         FatalError("Failed to retrieve \"HAS_VT\" atom\n");
     for (i = 0; i < xf86NumScreens; i++) {
-        ChangeWindowProperty(xf86ScrnToScreen(xf86Screens[i])->root,
-                             property_name, XA_INTEGER, 32,
-                             PropModeReplace, 1, &value, TRUE);
+        dixChangeWindowProperty(serverClient,
+                                xf86ScrnToScreen(xf86Screens[i])->root,
+                                property_name, XA_INTEGER, 32,
+                                PropModeReplace, 1, &value, TRUE);
     }
 }
 
diff --git a/include/property.h b/include/property.h
index e350513..be875e9 100644
--- a/include/property.h
+++ b/include/property.h
@@ -67,15 +67,6 @@ extern _X_EXPORT int dixChangeWindowProperty(ClientPtr pClient,
                                              void *value,
                                              Bool sendevent);
 
-extern _X_EXPORT int ChangeWindowProperty(WindowPtr pWin,
-                                          Atom property,
-                                          Atom type,
-                                          int format,
-                                          int mode,
-                                          unsigned long len,
-                                          void *value,
-                                          Bool sendevent);
-
 extern _X_EXPORT int DeleteProperty(ClientPtr /*client */ ,
                                     WindowPtr /*pWin */ ,
                                     Atom /*propName */ );
-- 
cgit v0.10.2

