diff -up ./nss/cmd/bltest/Makefile.iquote ./nss/cmd/bltest/Makefile
--- ./nss/cmd/bltest/Makefile.iquote	2016-06-20 10:11:28.000000000 -0700
+++ ./nss/cmd/bltest/Makefile	2016-06-24 15:04:36.742545266 -0700
@@ -45,6 +45,8 @@ include $(CORE_DEPTH)/coreconf/rules.mk
 # (6) Execute "component" rules. (OPTIONAL)                           #
 #######################################################################
 
+INCLUDES += -iquote $(DIST)/../public/nss
+INCLUDES += -iquote $(DIST)/../private/nss
 
 
 #######################################################################
diff -up ./nss/coreconf/location.mk.iquote ./nss/coreconf/location.mk
--- ./nss/coreconf/location.mk.iquote	2016-06-20 10:11:28.000000000 -0700
+++ ./nss/coreconf/location.mk	2016-06-24 15:04:36.820545264 -0700
@@ -45,6 +45,10 @@ endif
 
 ifdef NSS_INCLUDE_DIR
     INCLUDES += -I$(NSS_INCLUDE_DIR)
+    ifdef IN_TREE_FREEBL_HEADERS_FIRST
+        INCLUDES += -iquote $(DIST)/../public/nss
+        INCLUDES += -iquote $(DIST)/../private/nss
+    endif
 endif
 
 ifndef NSS_LIB_DIR
diff -up ./nss/lib/softoken/Makefile.iquote ./nss/lib/softoken/Makefile
--- ./nss/lib/softoken/Makefile.iquote	2016-06-20 10:11:28.000000000 -0700
+++ ./nss/lib/softoken/Makefile	2016-06-24 15:04:36.820545264 -0700
@@ -52,6 +52,9 @@ ifdef NSS_DISABLE_DBM
 DIRS=
 endif
 
+INCLUDES += -iquote $(DIST)/../public/nss
+INCLUDES += -iquote $(DIST)/../private/nss
+
 #######################################################################
 # (7) Execute "local" rules. (OPTIONAL).                              #
 #######################################################################
