diff -Nur libxcb/src/xcb_out.c libxcbn/src/xcb_out.c
--- libxcb/src/xcb_out.c	2015-07-12 22:17:06.000000000 +0800
+++ libxcbn/src/xcb_out.c	2015-10-19 13:28:42.204302478 +0800
@@ -215,7 +215,7 @@
         num_fds--;
     }
 #endif
-    close_fds(fds, num_fds);
+    //close_fds(fds, num_fds);
 }
 
 uint64_t xcb_send_request_with_fds64(xcb_connection_t *c, int flags, struct iovec *vector,
@@ -227,7 +227,7 @@
     enum workarounds workaround = WORKAROUND_NONE;
 
     if(c->has_error) {
-        close_fds(fds, num_fds);
+        //close_fds(fds, num_fds);
         return 0;
     }
 
@@ -248,7 +248,7 @@
             const xcb_query_extension_reply_t *extension = xcb_get_extension_data(c, req->ext);
             if(!(extension && extension->present))
             {
-                close_fds(fds, num_fds);
+                //close_fds(fds, num_fds);
                 _xcb_conn_shutdown(c, XCB_CONN_CLOSED_EXT_NOTSUPPORTED);
                 return 0;
             }
@@ -279,7 +279,7 @@
         }
         else if(longlen > xcb_get_maximum_request_length(c))
         {
-            close_fds(fds, num_fds);
+            //close_fds(fds, num_fds);
             _xcb_conn_shutdown(c, XCB_CONN_CLOSED_REQ_LEN_EXCEED);
             return 0; /* server can't take this; maybe need BIGREQUESTS? */
         }
@@ -361,7 +361,7 @@
     int fds[1] = { fd };
 
     if (c->has_error) {
-        close(fd);
+        //close(fd);
         return;
     }
     pthread_mutex_lock(&c->iolock);
