--- elfutils-0.160n/configure.ac	2014-10-25 10:20:22.821634927 +0000
+++ elfutils-0.160/configure.ac	2014-10-25 10:26:17.357607153 +0000
@@ -100,7 +100,7 @@
 
 AC_CACHE_CHECK([for -fgnu89-inline option to $CC], ac_cv_cc_gnu89_inline, [dnl
 old_CFLAGS="$CFLAGS"
-CFLAGS="$CFLAGS -fgnu89-inline -Werror"
+CFLAGS="$CFLAGS -fgnu89-inline"
 AC_COMPILE_IFELSE([AC_LANG_SOURCE([
 void foo (void)
 {
diff -Nur elfutils-0.160/config/eu.am elfutils-0.160n/config/eu.am
--- elfutils-0.160/config/eu.am	2014-10-25 10:34:33.311568301 +0000
+++ elfutils-0.160n/config/eu.am	2014-10-25 10:34:43.274567520 +0000
@@ -35,7 +35,6 @@
 DEFS = -D_GNU_SOURCE -DHAVE_CONFIG_H -DLOCALEDIR='"${localedir}"'
 AM_CPPFLAGS = -I. -I$(srcdir) -I$(top_srcdir)/lib -I..
 AM_CFLAGS = -std=gnu99 -Wall -Wshadow \
-	    $(if $($(*F)_no_Werror),,-Werror) \
 	    $(if $($(*F)_no_Wunused),,-Wunused $(WEXTRA)) \
 	    $(if $($(*F)_no_Wformat),-Wno-format,-Wformat=2) \
 	    $($(*F)_CFLAGS)
