diff -Nur efibootmgr-0.5.4/src/include/scsi_ioctls.h efibootmgr-0.5.4n/src/include/scsi_ioctls.h
--- efibootmgr-0.5.4/src/include/scsi_ioctls.h	2007-11-13 03:44:20.000000000 -0800
+++ efibootmgr-0.5.4n/src/include/scsi_ioctls.h	2014-10-24 09:45:59.571217794 -0800
@@ -34,8 +34,8 @@
 } Scsi_Idlun;
 
 
-inline int get_scsi_idlun(int fd, Scsi_Idlun *idlun);
-inline int get_scsi_pci(int fd, char *slot_name);
+int get_scsi_idlun(int fd, Scsi_Idlun *idlun);
+int get_scsi_pci(int fd, char *slot_name);
 int idlun_to_components (Scsi_Idlun *idlun,
 			 unsigned char *host,
 			 unsigned char *channel,
