diff -Nur doxygen-1.7.1/doc/Makefile doxygen-1.7.1n/doc/Makefile
--- doxygen-1.7.1/doc/Makefile	2010-06-25 17:08:35.000000000 +0800
+++ doxygen-1.7.1n/doc/Makefile	2011-01-10 18:02:23.000000000 +0800
@@ -43,7 +43,7 @@
 	@cp Makefile.latex ../latex/Makefile
 	@sed -e "s/\$$VERSION/$(VERSION)/g" doxygen_manual.tex >../latex/doxygen_manual.tex
 	@sed -e "s/\$$VERSION/$(VERSION)/g" doxygen.sty >../latex/doxygen.sty 
-	@epstopdf doxygen_logo.eps --outfile=../latex/doxygen_logo.pdf
+#	@epstopdf doxygen_logo.eps --outfile=../latex/doxygen_logo.pdf
 
 clean: 
 	rm -rf ../html ../latex *.bak 
diff -Nur doxygen-1.7.1/doc/Makefile.in doxygen-1.7.1new/doc/Makefile.in
--- doxygen-1.7.1/doc/Makefile.in	2008-12-06 21:16:23.000000000 +0800
+++ doxygen-1.7.1new/doc/Makefile.in	2011-01-10 18:06:05.000000000 +0800
@@ -23,7 +23,7 @@
 	@cp Makefile.latex ../latex/Makefile
 	@sed -e "s/\$$VERSION/$(VERSION)/g" doxygen_manual.tex >../latex/doxygen_manual.tex
 	@sed -e "s/\$$VERSION/$(VERSION)/g" doxygen.sty >../latex/doxygen.sty 
-	@epstopdf doxygen_logo.eps --outfile=../latex/doxygen_logo.pdf
+#	@epstopdf doxygen_logo.eps --outfile=../latex/doxygen_logo.pdf
 
 clean: 
 	rm -rf ../html ../latex *.bak 
