diff -Nur mesa/src/mesa/drivers/dri/i965/brw_context.c mesan/src/mesa/drivers/dri/i965/brw_context.c
--- mesa/src/mesa/drivers/dri/i965/brw_context.c	2015-08-12 13:06:46.000000000 +0000
+++ mesan/src/mesa/drivers/dri/i965/brw_context.c	2015-08-13 10:40:52.163958379 +0000
@@ -1402,8 +1402,8 @@
               buffer->name, buffer->attachment,
               buffer->cpp, buffer->pitch);
    }
-
-   intel_miptree_release(&rb->mt);
+   //https://bugs.freedesktop.org/show_bug.cgi?id=86281
+   //intel_miptree_release(&rb->mt);
    bo = drm_intel_bo_gem_create_from_name(brw->bufmgr, buffer_name,
                                           buffer->name);
    if (!bo) {
