diff -Nur mesa/src/mesa/drivers/dri/i965/brw_meta_fast_clear.c mesanew/src/mesa/drivers/dri/i965/brw_meta_fast_clear.c
--- mesa/src/mesa/drivers/dri/i965/brw_meta_fast_clear.c	2015-07-15 17:45:00.171970341 +0000
+++ mesanew/src/mesa/drivers/dri/i965/brw_meta_fast_clear.c	2015-07-15 17:45:19.250970000 +0000
@@ -499,7 +499,7 @@
             clear_type = REP_CLEAR;
 
       if (irb->mt == NULL)
-         clear_type = REP_CLEAR;
+         clear_type = PLAIN_CLEAR;
 
       switch (clear_type) {
       case FAST_CLEAR:
